### ISSUE TITLE
[System.Text.Json] JsonInclude does not support JsonExtensionData

### ISSUE THREAD

**Issue by RamjotSingh** [dotnet/runtime#54192]
JsonExtensionData allows having an open ended dictionary to hold overflow items. JsonInclude allow properties with missing getters and setters to be serialized (or deserialized). However it looks like they do not work together.

#### Works
```cs
         /// <summary>
        /// Gets or sets the dynamic properties.
        /// </summary>
        [System.Text.Json.Serialization.JsonExtensionData]
        public IDictionary<string, object> Properties { get; set; } = new Dictionary<string, object>();
```

#### Does not work (without JsonInclude)
```cs
        /// <summary>
        /// Gets or sets the dynamic properties.
        /// </summary>
        [System.Text.Json.Serialization.JsonExtensionData]
        public IDictionary<string, object> Properties { get; } = new Dictionary<string, object>();
```

#### Does not work (without JsonInclude)
```cs
        /// <summary>
        /// Gets or sets the dynamic properties.
        /// </summary>
        [System.Text.Json.Serialization.JsonExtensionData]
        [JsonInclude]
        public IDictionary<string, object> Properties { get; } = new Dictionary<string, object>();
```

In general I believe collections where init is not needed (like the one above) by library should not even need `JsonInclude`. It is a common scenario (and one of the StyleCop rules) to not allow setting a collection to prevent them going `null`'.

**Comment by eiriktsarpalis** [comment]
Hi @RamjotSingh, according to the [`JsonIncludeAttribute` documentation](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonincludeattribute?view=net-5.0) the attribute is applied to enable access to non-public getters and setters. However, your example does not define any setters. Once I change your example to the following:
```csharp
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

var poco = new Poco();
poco.Properties.Add("key", 42);

string json = JsonSerializer.Serialize(poco);
var deserialized = JsonSerializer.Deserialize<Poco>(json);
Console.WriteLine(deserialized!.Properties.Count); // prints 0, uncomment attribute to change result to 1.

public class Poco
{
    [JsonExtensionData]
    //[JsonInclude]
    public IDictionary<string, object> Properties { get; private set; } = new Dictionary<string, object>();
}
```
`JsonIncludeAttribute` works as expected. But perhaps I'm misunderstanding your issue, could you please share a standalone minimal reproduction indicating the precise expected vs actual behavior?

**Comment by RamjotSingh** [comment]
Adding private set; (and JsonInclude) makes it work. But the original intent is to not have setter at all. Since the init is already done.

**Comment by RamjotSingh** [comment]
Not having setter is similar to init; kind. For collections for example you do not want people to override and accidently set it to null. By removing setter you ensure that. Hence in such cases, having setters is bad.

**Comment by eiriktsarpalis** [comment]
> Not having setter is similar to init; kind.

That's not accurate, in fact `init` is a language-level feature which still results in a setter method being emitted in the IL, see for example:

https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIQCGAzpQQN44GNGFgB2wBAgnQQOYCmwANwE2YYQQD0AKgKVBI6ZIIBfHCqA=

It is not possible for a serializer to populate a property without a setter. You might be able in some cases to deduce the field backing the property and write to that, however that doesn't always work and is considered poor practice. In many cases there is no backing field as such, so populating the property is impossible outright:
```csharp
public class Foo
{
    public int X => System.Random().Next();
}
```

**Comment by RamjotSingh** [comment]
Fair enough. If it throws a wrench into how library deduces things I think it is fair to ask for a setter.

**Comment by ericstj** [comment]
Do we have precedent in JSON for allowing it to modify existing collections rather than creating them and setting them?    @layomia is the expert here.  I seem to remember us having a strong opinion about this.

**Comment by ericstj** [comment]
I think I found the answer to this myself, all our converters today create new collections and set them, so this is inline with current precedent in the library.
https://github.com/dotnet/runtime/blob/36e0432547d7d734ef23fce40e8c278ef80c34c8/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ICollectionOfTConverter.cs#L37
https://github.com/dotnet/runtime/blob/36e0432547d7d734ef23fce40e8c278ef80c34c8/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryOfTKeyTValueConverter.cs#L39
dotnet/runtime.NET is a cross-platform runtime for cloud, mobile, desktop, and IoT apps. - dotnet/runtimegithub.com
